### PR TITLE
Reward relayers for dispatching messages

### PR DESCRIPTION
**Pull request by svyatonik**
related #316
follow up of #339 (II)

This PR introduces scheme which rewards relayers that have delivered messages to the target chain. There are no any transfers - reward logic is hidden behind `MessageDeliveryAndDispatchPayment` trait, so both "the winner takes all" and election scheme from #316 can be applied.

Main change is that inbound lane state now holds set of relayers which have delivered last messages. Since proof-of-delivery is that state, the set is included into the "delivery race" and is relayed back to the source chain.

The main drawback of this approach is that if there are too may relayed messages, this set may grow indefinitely. This is solved by introducing additional config param `MaxUnconfirmedMessagesAtInboundLane`. If relayers are not confirming delivery back to the source chain for too long, and there are `MaxUnconfirmedMessagesAtInboundLane` unconfirmed messages in the inbound lane, then it stops accepting new messages && waits until proof-of-outbound-lane-state will be included into the messages-proof. When updated outbound lane state is received, all rewarded relayers ids are removed from the relayers set. This would require minimal changes into "delivery race" logic of the relayer - I'll add those changes in separate PR.

### CODE REVIEW

File: modules/message-lane/src/mock.rs
Comment (HCastano) [inline modules/message-lane/src/mock.rs:224]: Is using functionality from the `unhashed` module only useful for testing, or does it have applications beyond that?

**Review by HCastano — APPROVED** [review]
Looks fine, but I'd wait until Tomek takes a look before merging

File: modules/message-lane/src/lib.rs
Comment (svyatonik) [reply modules/message-lane/src/lib.rs:300]: Can remove. But changing to something that'll work in release isn't desired, because this `group_by` is actually happens in the same pallet function.

File: primitives/message-lane/src/target_chain.rs
Comment (svyatonik) [reply primitives/message-lane/src/target_chain.rs:26]: When you relay messages to the target chain, they aren't simply encoded messages. Instead we relay storage-proof of these messages. `ProvedMessages` are the messages that are proved by that proof (of proved to be in that proof). If you have any other identifier idea - please suggest.

File: modules/message-lane/src/mock.rs
Comment (svyatonik) [reply modules/message-lane/src/mock.rs:224]: It is also used for some keys that are shared by all runtimes (aka 'well_known_keys') - i.e. ":code", ":max_heap_pages", ...

File: primitives/message-lane/src/lib.rs
Comment (svyatonik) [reply primitives/message-lane/src/lib.rs:76]: I've added another small comment to the code

File: modules/message-lane/src/inbound_lane.rs
Comment (tomusdrw) [inline modules/message-lane/src/inbound_lane.rs:90]:
```suggestion
		// if there are more unconfirmed messages than we may accept, reject this message
```

File: primitives/message-lane/src/target_chain.rs
Comment (tomusdrw) [inline primitives/message-lane/src/target_chain.rs:86]: We don't verify here that `messages` are actually coming from the same `lane`. While in theory it shouldn't happen, having this data duplicated may just lead to mistakes. Maybe let's change `Message/DispatchMessage` to only contain `nonce` (instead of entire `key`) and add the lane information only during actual `receive_message`?

File: modules/message-lane/src/lib.rs
Comment (tomusdrw) [inline modules/message-lane/src/lib.rs:304]: Why don't we reject the entire transaction here? Previously we would get `false` here in case there was a nonce mismatch, now it may also happen if there is simply too many unconfirmed deliveries. I feel this is not something that is super desirable, but at the same time can't find any significant problems with the current approach.

File: modules/message-lane/src/lib.rs
Comment (tomusdrw) [inline modules/message-lane/src/lib.rs:73]: While this works I really doubt it's the right abstraction. It not only limits the number of "unconfirmed messages", but it actually also limits the number of messages you can dispatch in a single transaction (even though it's a single relayer).

A really simple optimisation would be to store `VecDeque<(StartNonce, Option<EndNonce>, Relayer)>` instead of duplicating the relayer for each nonce it delivered. We really expect the relayers to dispatch multiple messages at once, so IMHO it makes sense to optimize that. Thanks to this, if a relayer delivers a batch, it will only occupy one slot in that collection, so the `MaxUncofirmedMessagesAtInboundeLane` wouldn't restrict the message delivery that much.

File: modules/message-lane/src/lib.rs
Comment (tomusdrw) [reply modules/message-lane/src/lib.rs:304]: I was initially assuming that the delivery proof is going to be simply the transaction proof, but I've realised is that this makes the `verify_messages_delivery_proof` function harder to implement and potentially increase the size of the proof unnecessarily.
It seems that you assume that the delivery proof is actually going to be simply a storage proof, we don't necessarily care that much.

File: modules/message-lane/src/lib.rs
Comment (tomusdrw) [inline modules/message-lane/src/lib.rs:322]: We might consider incentivising this separately. Currently it's assumed that the relayer who delivered the messages is interested to get rewards. However, we should make sure that other relayers are not going to get blocked by him (which currently is the case).
We might split the total fee between the relayers who delivered a message and the submitter of the delivery proof. CCing #316 here.

**Review by tomusdrw — COMMENTED** [review]
Looks good, few concerns in comments.

File: modules/message-lane/src/inbound_lane.rs
Comment (svyatonik) [reply modules/message-lane/src/inbound_lane.rs]: Yes, it was original idea from #339. There were multiple problems with this approach. Can't remember all, since a week has already passed, but here are some examples:
1) if you have some sort of queue (store relayers of last 1024 messages), it'll mean that every confirmation transaction after 1024 messages is guaranteed to have these additional >32kb bytes, unless you'll find a way to prove that reward has been paid (i.e. what is presented in this PR);
2) if you're going to introduce periods in blocks instead (store relayers + number of messages that they have relayed in last 1024 blocks), it'll require additional storage entry + separate transaction type to deliver that. Also - size of confirmation transaction may grow indefinitely (because we don't limit number of messages/1024 blocks);
3) since confirmation transaction requires source chain to know about target block' finality, it means that reward requires: (3.a) GRANDPA authorities to finalize target chain blocks in timely fashion (3.b) header relayers to relay target chain blocks and finality to source chain and (3.c) message relayer to submit delivery transaction. So there are too many other players (GRANDPA authorities, headers relayers, other relayers, nodes txpools) that may block claiming reward;
4) if you're going to introduce claim-reward transaction for relayers that have missed their opportunity (i.e. GRANDPA authorities have failed to finalize target chain blocks in timely fashion), it would mean that you'll need some way to mark reward as claimed (i.e. store unclaimed rewards separately OR change how pruning works).

File: primitives/message-lane/src/source_chain.rs
Comment (svyatonik) [reply primitives/message-lane/src/source_chain.rs:90]: I think we may remove it.

File: primitives/message-lane/src/target_chain.rs
Comment (svyatonik) [reply primitives/message-lane/src/target_chain.rs:86]: > We don't verify here that messages are actually coming from the same lane. While in theory it shouldn't happen, having this data duplicated may just lead to mistakes. Maybe let's change Message/DispatchMessage to only contain nonce (instead of entire key) and add the lane information only during actual receive_message?

`verify_messages_proof` could return anything it wants - it can e.g. replace all messages payloads with `BalancesCall::transfer(svyatonik, 1 DOT)`. So you just need to trust it.

We can't remove `lane` from the MessageKey, because in this way relayer can just submit messages from Lane1 to Lane2.

File: modules/message-lane/src/lib.rs
Comment (svyatonik) [reply modules/message-lane/src/lib.rs:304]: > Why don't we reject the entire transaction here? Previously we would get false here in case there was a nonce mismatch, now it may also happen if there is simply too many unconfirmed deliveries. I feel this is not something that is super desirable, but at the same time can't find any significant problems with the current approach.

Because, as it was previously, transactions may overlap. Tx1[msg1, msg2, msg3] from relayer1 may be mined earler than Tx2[msg1, msg2, msg3, msg4] from relayer2, but it still may bring useful data (msg4). Relayer has all means to determine when it need to include `proof-of(outbound_lane.state)` into delivery transaction.

File: modules/message-lane/src/lib.rs
Comment (svyatonik) [reply modules/message-lane/src/lib.rs:73]: > While this works I really doubt it's the right abstraction. It not only limits the number of "unconfirmed messages", but it actually also limits the number of messages you can dispatch in a single transaction (even though it's a single relayer).

If you care about possible number of mesages in a single transaction, make this at least `MaxBlockWeight / MaxExtrinsicWeight`. But in practice it should be much larger, or we will be blocked by GRANDPA authorities + headers relays.

> A really simple optimisation would be to store VecDeque<(StartNonce, Option<EndNonce>, Relayer)> instead of duplicating the relayer for each nonce it delivered. We really expect the relayers to dispatch multiple messages at once, so IMHO it makes sense to optimize that. Thanks to this, if a relayer delivers a batch, it will only occupy one slot in that collection, so the MaxUncofirmedMessagesAtInboundeLane wouldn't restrict the message delivery that much.

Yes - this may be done, but since it is only a potential optimization - let's file a separate issue for that.

File: modules/message-lane/src/lib.rs
Comment (svyatonik) [reply modules/message-lane/src/lib.rs:322]: > We might consider incentivising this separately. Currently it's assumed that the relayer who delivered the messages is interested to get rewards. However, we should make sure that other relayers are not going to get blocked by him (which currently is the case).

Other relayers always have an opportunity to unblock themselves. But I agree that incentivisation hook may be done here. Actually, probably adding `confirmation_transaction_submitter` to `pay_relayer_reward()` would be enough.

File: modules/message-lane/src/lib.rs
Comment (svyatonik) [reply modules/message-lane/src/lib.rs:73]: > MaxBlockWeight / MaxExtrinsicWeight

This is obviously wrong, sorry. Should be something like `MaxBlockWeight / Avg(MaxExtrinsicWeight)`. Nevertheless, it is still relevant when talking about messages/tx.

Comment (tomusdrw) [reply modules/message-lane/src/lib.rs:73]: Extracted as #406

File: primitives/message-lane/src/target_chain.rs
Comment (tomusdrw) [reply primitives/message-lane/src/target_chain.rs:86]: My point was that this information is now duplicated. We already have lane id in the returned map, why do we need it in the message itself?

Comment (svyatonik) [reply primitives/message-lane/src/target_chain.rs:86]: Ok - will remove from trait

Comment (svyatonik) [reply primitives/message-lane/src/target_chain.rs]: done

File: modules/message-lane/src/lib.rs
Comment (tomusdrw) [inline modules/message-lane/src/lib.rs:405]: Couldn't we use like a `RefCell` to avoid adding `&mut` here?

**Review by tomusdrw — APPROVED** [review]
lgtm!

**Pull request merged**
